File: alsh_mips/src/python/lsh_wrapper.py
# -*- coding: utf-8 -*-

"""
# Locality Sensitive Hashing for MIPS

## Reference:

- LSH:
[Locality-Sensitive Hashing Scheme Based on p-Stable Distributions](http://www.cs.princeton.edu/courses/archive/spr05/cos598E/bib/p253-datar.pdf)

- L2-ALSH:
[Asymmetric LSH (ALSH) for Sublinear Time Maximum Inner Product Search (MIPS)](https://papers.nips.cc/paper/5329-asymmetric-lsh-alsh-for-sublinear-time-maximum-inner-product-search-mips.pdf)

- Sign-ALSH(Cosine):
[Improved Asymmetric Locality Sensitive Hashing (ALSH) for Maximum Inner Product Search (MIPS)](http://auai.org/uai2015/proceedings/papers/96.pdf)

- Simple-ALSH(Cosine):
[On Symmetric and Asymmetric LSHs for Inner Product Search](https://arxiv.org/abs/1410.5518)

"""

import random
from abc import ABCMeta, abstractmethod
from collections import defaultdict
from operator import itemgetter
from global_mgr import gol

from lsh import *

class LshWrapper:
    'LSH Wrapper'

    # lsh_type: lsh hash func type: in 'l2' 'cosine'
    # r: random float data
    # d: data vector size
    # k: number of hash func for each hashtable. default 2
    # L: number of hash tables for each hash type: default 2
    def __init__(self, lsh_type, d, r = 1.0, k = 2, L = 2):
        self.type = lsh_type
        self.d = d
        self.r = r
        self.k = k
        self.L = 0
        self.hash_tables = []
        self.resize(L)

        if True == gol.get_value('DEBUG'):
            print 'LshWrapper init:\ttype: ' + str(self.type) + '\td: ' + str(self.d) + '\tr: ' + str(self.r) + '\tk: ' + str(self.k) + '\tL: ' + str(self.L)

    def __get_hash_class__(self):
        try:
            if 'l2' == self.type:
                return L2Lsh
            elif 'cosine' == self.type:
                return CosineLsh
            else:
                raise ValueError
        except ValueError:
            print "LshWrapper type error: " + self.type
            return

    def __creat_ht__(self):
        try:
            if 'l2' == self.type:
                return L2Lsh(self.r, self.d)
            elif 'cosine' == self.type:
                return CosineLsh(self.d)
            else:
                raise ValueError
        except ValueError:
            print "LshWrapper type error: " + self.type
            return

    def resize(self, L):
        if True == gol.get_value('DEBUG'):
            print 'LshWrapper resize:\tnew L: ' + str(L) + '\tL: ' + str(self.L) + '\tk: ' + str(self.k)

        # shrink the number of hash tables to be used
        if L < self.L:
            self.hash_tables = self.hash_tables[:L]
        else:
            # initialise a new hash table for each hash function
            hash_funcs = [[self.__creat_ht__() for h in xrange(self.k)] for l in xrange(self.L, L)]
            self.hash_tables.extend([(g, defaultdict(lambda:[])) for g in hash_funcs])

            if True == gol.get_value('DEBUG'):
                print 'resize wrapper hashtable: '
                for ht, ct in self.hash_tables:
                    print 'ht: ' + str(type(ht)) + '\tct: ' + str(type(ct))
                    print ht
                    print ct
        self.L = L

    def hash(self, ht, data):
        #  used for combine
        return self.__get_hash_class__().combine([h.hash(data) for h in ht])

    def index(self, datas):
        # index the supplied datas
        self.datas = datas
        for ht, ct in self.hash_tables:
            for ix, p in enumerate(self.datas):
                ct[self.hash(ht, p)].append(ix)
        # reset stats
        self.tot_touched = 0
        self.num_queries = 0

        if True == gol.get_value('DEBUG'):
            print 'index wrapper hashtable: '
            for ht, ct in self.hash_tables:
                print 'ht: ' + str(type(ht)) + '\tct: ' + str(type(ct))
                print ht
                print ct

    def query(self, q, metric, max_results = 1):
        """
        triple_l = 3 * self.L
        if max_results > triple_l:
            max_results = triple_l
        elif
        """
        if 0 == max_results:
            max_results = 1

        # find the max_results closest indexed datas to q according to the supplied metric
        candidates = set()
        for ht, ct in self.hash_tables:
            matches = ct.get(self.hash(ht, q), [])
            candidates.update(matches)

        # update stats
        self.tot_touched += len(candidates)
        self.num_queries += 1

        # rerank candidates
        candidates = [(ix, metric(q, self.datas[ix])) for ix in candidates]
        candidates.sort(key=itemgetter(1))
        return candidates[:max_results]

    def get_avg_touched(self):
        # mean number of candidates inspected per query
        return self.tot_touched / self.num_queries

    def distance(self, u, v):
        return __get_hash_class__.distance(u, v)

File: alsh_mips/src/cpp/lsh_wrapper.cpp
#include "lsh_wrapper.h"
#include <iostream>
#include <algorithm>

using namespace std;

LshWrapper::LshWrapper() {
}

LshWrapper::~LshWrapper() {
}

int LshWrapper::init(LshType type, int d, double r , int k, int L) {
  release_memory();

  type_ = type;
  d_ = d;
  r_ = r;
  k_ = k;
  L_ = L;

  create_hash();

  for (int i = 0; i < L; ++i) {
    std::multimap<std::string, long> *m = new std::multimap<std::string, long>();
    hash_map_.push_back(m);
  }

  return 0;
}

int LshWrapper::create_hash() {
  for (int i = 0; i < L_; ++i) {
    std::vector<Hash *> h_vec;
    for (int j = 0; j < k_; ++j) {
      Hash *h = NULL;
      if (type_ == L2_LSH) {
        h = new L2Lsh(r_, d_);
        // h->init(r_, d_);
      } else if (type_ == CONSINE_LSH) {
        h = new ConsineLsh(r_, d_);
        // h->init(r_, d_);
      } else {
        cout<<"error LSH type !"<<endl;
      }

      h_vec.push_back(h);
    }

    hashes_func_.push_back(h_vec);
  }
  return 0;
}
void LshWrapper::release_memory() {
  for (int i = 0; i < hashes_func_.size(); ++i) {
    for (int j = 0; j < hashes_func_[i].size(); ++j) {
      delete hashes_func_[i][j];
      hashes_func_[i][j] = NULL;
    }

    hashes_func_[i].clear();
  }

  hashes_func_.clear();
  for (int i = 0; i < hash_map_.size(); ++i) {
    delete hash_map_[i];
     hash_map_[i] = NULL;
  }

  hash_map_.clear();
}

int LshWrapper::resize(int L) {
  return 0;
}

std::string LshWrapper::hash(const std::vector<Hash *> &hash_vec, const VectorXd &data) {
  std::vector<int> hash_val;
  int vec_size = hash_vec.size();
  for (int i = 0; i < vec_size; ++i) {
    hash_val.push_back(hash_vec[i]->hash(data));
  }

  if (type_ == L2_LSH) {
    return hash_combine(hash_val);
  } else if (type_ == CONSINE_LSH) {
    return cosine_hash_combine(hash_val);
  } else {
    cout<<"error LSH type !"<<endl;
    return "";
  }
}

int LshWrapper::index(const MatrixXd &datas) {
  this->datas = datas;
  std::vector<std::vector<Hash *> >::const_iterator hash_iter;
  std::vector<std::multimap<std::string, long> * >::const_iterator vec_iter;
  for (hash_iter = hashes_func_.begin(), vec_iter = hash_map_.begin();
      hash_iter != hashes_func_.end() && vec_iter != hash_map_.end();
      vec_iter++, hash_iter++) {
    int row_size = datas.rows();
    for (int i = 0; i < row_size; ++i) {
      std::multimap<std::string, long> *cur_map = *vec_iter;
      cur_map->insert(std::pair<std::string, long>(this->hash(*hash_iter, datas.row(i)), i));
    }
  }

  tot_touched_ = 0.0;
  num_queries_ = 0;
  return 0;
}

VectorXd LshWrapper::query(const VectorXd &q, metric_func metric, int max_result) {
  VectorXd ret_v(max_result);

  #if 0
  triple_l = 3 * self.L
  if max_results > triple_l:
      max_results = triple_l
  elif
  #endif

  if (0 == max_result) {
    max_result = 1;
  }

  // find the max_results closest indexed datas to q according to the supplied metric
  multimap<double,long> candidates;  //double:distance    long:index
  vector<long> candidate_index;
  std::vector<std::vector<Hash *> >::const_iterator hash_iter;
  std::vector<std::multimap<std::string, long> * >::const_iterator vec_iter;
  typedef multimap<string, long>::iterator multi_iter;
  pair<multi_iter, multi_iter> iter;
  string footprint_q;
  double distance;

  // cout<<"max_result = "<<max_result<<endl;
  // cout<<"hashes_func_size = "<<hashes_func_.size()<<endl;
  for (hash_iter = hashes_func_.begin(), vec_iter = hash_map_.begin();
      hash_iter != hashes_func_.end() && vec_iter != hash_map_.end();
      vec_iter++, hash_iter++) {
    footprint_q = this->hash(*hash_iter, q);
    // cout<<"footprint_q= "<<footprint_q<<endl;
    std::multimap<std::string, long> *cur_map = *vec_iter;
    // for(std::multimap<std::string, long>::iterator it=(*cur_map).begin();it!=(*cur_map).end();it++)
    //  cout<<it->first<<" "<<it->second<<endl;
    iter = cur_map->equal_range(footprint_q);
    for(multi_iter k = iter.first; k != iter.second; k++) {
      distance = metric(q, datas.row(k->second));
      if(find(candidate_index.begin(),candidate_index.end(),k->second) == candidate_index.end()) {
        candidates.insert(pair<double,long>(distance,k->second));
        candidate_index.push_back(k->second);
      }
    }
  }

  // for(multimap<double,long>::iterator it = candidates.begin();it!= candidates.end();it++) {
  //  cout<<"distance "<<it->first<<" index "<<it->second<<endl;
  // }

  // update stats
  tot_touched_ += candidates.size();
  num_queries_ += 1;
  // cout<<"candidates.size "<<candidates.size()<<endl;
  multimap<double,long>::iterator iter_sort=candidates.begin();
  for(int i=0;i<max_result&&iter_sort!=candidates.end();i++,iter_sort++) {
    ret_v(i) = iter_sort->second;
  }

  return ret_v;
}

double LshWrapper::get_avg_touched() {
  return double(tot_touched_ / num_queries_);
}
